### ISSUE TITLE
Proposal: Add mechanism to handle curcular references when Serializing

### ISSUE THREAD

**Issue by Jozkee** [dotnet/corefx#41002]
# Rationale and Usage

Currently there is no mechanism to prevent infinite looping when an object that contains a circular reference is serialized using `JsonSerializer.Serialize`. 
You can find more information about how this works in Json.Net reading this [gist](https://gist.github.com/Jozkee/e96cfde875b4773086fadf3aa1017a10).

Issues related:
* https://github.com/dotnet/corefx/issues/38579
* https://github.com/dotnet/corefx/issues/37786
* https://github.com/dotnet/corefx/issues/40045

Take this for example:

Given the following class:
```
class Employee { 
    string Name { get; set; }
    Employee Manager { get; set; }
    List<Employee> Subordinates { get; set; }
}
```
And having the following code:
```
var Angela = new Employee();
var Bob = new Emploee();

Angela.Subordinates  new List<Employee> { Bob };
Bob.Manager = Angela;

string json = JsonSerializer.Serialize(Angela);
```

You will find that the above code throws by not being able to keep up when MaxDepth is reached:
https://github.com/dotnet/corefx/blob/50e8db3cabbfb6ad0fc0702b3db25491a95c72a5/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.cs#L61-L64

This is the only thing we have before hitting an stack overflow. 

# Proposed API
```
namespace System.Text.Json
{
  public enum ReferenceHandling: byte
  {
      Error = (byte)0,
      Ignore = (byte)1,
      Preserve = (byte)2,
  }

  public sealed partial class JsonSerializerOptions
  {
      ReferenceHandling ReferenceHandling { get { throw null; } set { } }
  }
}

namespace System.Text.Json.Serialization
{
  [AttributeUsage(AttributeTargets.Property, AllowMultiple = false)]
  public sealed class JsonReferenceHandlingAttribute : JsonAttribute
  {
      public JsonReferenceHandlingAttribute(ReferenceHandling handling)
      {
          Handling = handling;
      }
      public ReferenceHandling Handling { get; }
  }
}
```

# Examples
Using the previous `Employee` class and serializing Angela with `ReferenceHandling.Ignore` option:
```
var Angela = new Employee { Name = "Angela" };
var Bob = new Employee { Name = "Bob" };

Angela.Subordinates  new List<Employee> { Bob };
Bob.Manager = Angela;

var options = new JsonSerializerOptions { 
    ReferenceHandling = ReferenceHandling.Ignore 
};

string json = JsonSerializer.Serialize(Angela, options);
Console.WriteLine(json);
```
Here's the output:
```
{
  "Name": "Angela",
  "Subordinates": [
    {
      "Name": "Bob"
    }
  ]
}
```

You can also deal with the loop using `ReferenceHandling.Preserve` option:
```
var Angela = new Employee { Name = "Angela" };
var Bob = new Employee { Name = "Bob" };

Angela.Subordinates  new List<Employee> { Bob };
Bob.Manager = Angela;

var options = new JsonSerializerOptions { 
    ReferenceHandling = ReferenceHandling.Preserve 
};

string json = JsonSerializer.Serialize(Angela, options);
Console.WriteLine(json);
```

And here's the output:
```
{
  "$id": "1"
  "Name": "Angela",
  "Subordinates": {
    "$id": "2",
    "values": [
      {
        "$id": "3"
        "Name": "Bob",
        "Manager": {
            "$ref": "1"
        }
      }
    ]
  }
}
```
Note how it also replaces arrays for objects in order to add `$id` property.

Another option would be to add `ReferenceHandling.Preserve` attribute on top of the property:
```

private class Employee
{
    public string Name { get; set; }

    [JsonReferenceHandling(ReferenceHandling.Preserve)]
    public Employee Manager { get; set; }

    public List<Employee> Subordinates { get; set; }
}

var Angela = new Employee { Name = "Angela" };
var Bob = new Employee { Name = "Bob" };

Angela.Subordinates  new List<Employee> { Bob };
Bob.Manager = Angela;

var options = new JsonSerializerOptions { 
    //Even though you are setting Ignore option, Manager will use Preserve
    ReferenceHandling = ReferenceHandling.Ignore 
};

string json = JsonSerializer.Serialize(Angela, options);
Console.WriteLine(json);
```
Output:
```
{
  "Name": "Angela",
  "Subordinates": [
    {
      "Name": "Bob",
      "Manager": {
        "$id": "1",
        "Name": "Angela",
        "Subordinates": [
          {
            "Name": "Bob",
            "Manager": {
              "$ref": "1"
            }
          }
        ]
      }
    }
  ]
}
```
Notice it goes this deep because we are just preserving `Manager` property.

# Details
* Default option is Error.
* This API is currently available on Json.Net therefore most of the behavior will be similar to it.
* `ReferenceLoopHandling` and `PreserveReferenceHandling` options are being merged into the proposed enum `ReferenceHandling`.
* `ReferenceLoopHandling.Serialize` option will not be ported/included.
* `PreserveReferenceHandling.Object` and `Array` options will not be ported/included, just `All`.
* We will keep the same schema of metadata properties appended to the complex types in order to offer campatibility with Json.Net. 
* In order to be able to deserialize back a Json string serialized using `ReferenceHandling.Preserve` option, we will need to extend `JsonSerializer.Deserialize` so it can read the metadata tags (`$id` and `$ref`).
* I am adding the perserve mechanism within a property or class attribute as well as an option in `JsonSerializerOptions`, thing that is not on Json.Net.
* You can find more information about how this works in Json.Net reading the following gist: https://gist.github.com/Jozkee/e96cfde875b4773086fadf3aa1017a10

# Open Questions
* Do these APIs should exist in the namespaces that I suggested?
* The merge of both Json.Net options means losing the [granularity](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_PreserveReferencesHandling.htm) of `PreserveReferenceHandling`, this is just my opinion but I don't see why someone would opt-in for apply this to only object OR only arrays, so I think it is a cheap trade-off of merging both options.
* The merge also forces that we include `PreserveReferenceHandling` in the attribute level, thing that does not exist in Json.Net.
* Since a property with `JsonConverter` is forced to have a `ClassType` of `Value` or `Unknown`, you cannot combine it with `JsonReferenceHandling` which is only meant for complex types `Object` and `Array`. I think we can adjust code so it can correctly handle converted values but consider this for the trade-off.
* We were discussing to keep current validation so users can opt-in for not use this option if they want to keep performance. Is it ok to remove [current validation](https://github.com/dotnet/corefx/blob/50e8db3cabbfb6ad0fc0702b3db25491a95c72a5/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.cs#L61-L64) if the implementation involves some dictionary/set that provides O(1) runtime and does not impact performance?
* Also consider these scenarios:
  * This person is asking for a feature that disables `PreserveRefHandling` feature in the reader, should we include that?: https://stackoverflow.com/questions/22299390/can-not-deserialize-json-containing-ref-keys

**Comment by Jozkee** [comment]
@ahsonkhan @JamesNK thoughts?

**Comment by JamesNK** [comment]
Consider changing "values" in the array metadata object to "$values" to indicate it isn't a real property.

```json
{
  "$id": "1"
  "Name": "Angela",
  "Subordinates": {
    "$id": "2",
    "values": [
      {
        "$id": "3"
        "Name": "Bob",
        "Manager": {
            "$ref": "1"
        }
      }
    ]
  }
}
```

Instead of a dedicated property, what about adding ReferenceHandling as a property to `JsonPropertyAttribute` (is that the right name? I don't recall off the top of my head).
```cs
[JsonReferenceHandling(ReferenceHandling.Preserve)]
```

**Comment by Jozkee** [comment]
> Consider changing "values" in the array metadata object to "$values" to indicate it isn't a real property.

Agree.

> Instead of a dedicated property, what about adding ReferenceHandling as a property to JsonPropertyAttribute (is that the right name? I don't recall off the top of my head).

`JsonPropertyAttribute` does not exist in `System.Text.Json` as far as I know.

**Comment by ahsonkhan** [comment]
> `JsonPropertyAttribute` does not exist in `System.Text.Json` as far as I know.

Correct, we have `JsonPropertyNameAttribute`
https://github.com/dotnet/corefx/blob/b41b09eadd5eb3f5575845a982f2c177c37f7ce7/src/System.Text.Json/ref/System.Text.Json.cs#L748-L753

cc @steveharter

**Comment by scalablecory** [comment]
Is this proposal compatible with Json.Net?

**Comment by scalablecory** [comment]
```
ReferenceHandling: byte
```

Why `byte`?

**Comment by JamesNK** [comment]
What determines if objects are equal? object.ReferenceEquals or object.Equals? Newtonsoft.Json uses object.Equals.

**Comment by JamesNK** [comment]
What happens if there are additional properties in a JSON reference object? (has `$ref`) What about if $ref isn't the first property in an object?

What happens if `$id` isn't the first property?

**Comment by Jozkee** [comment]
> Is this proposal compatible with Json.Net?

Yes for the metadata semantics that I am trying to follow; no for a feature parity, however we are not tied to follow the first one, just come up with what is already there since I have not found value in going on a different direction.

> Why byte?

I was following existing enums in the namespace that also inherits from byte, thing that is made to reduce their size in the `Utf8JsonReader` struct. but since the Serializer is a class, you are right, there is no much value in changing the enum's default value, I will fix that. Thanks.

**Comment by Jozkee** [comment]
> What determines if objects are equal? object.ReferenceEquals or object.Equals? Newtonsoft.Json uses object.Equals.

Currently the implementation relies on default equality of `Hashset`/`Dictionary`, which uses the `Object.GetHashCode()` and `Object.Equals(Object obj)`.
If someone wants to provide its own equals mechanism, he can override these methods and/or implement `IEquatable` on its class.

> What happens if there are additional properties in a JSON reference object? (has $ref) What about if $ref isn't the first property in an object?

Json.Net throws when finds that there are properties other than $ref in a JSON reference object, I was planning on sticking to that behavior. 
I see no point in allowing other properties to co-exist with a $ref, unless the user relies on the $ref property for another purpose; for that maybe in a future we could add a mechanism to disable this functionality similar to [`MetadataPropertyHandling.Ignore`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_MetadataPropertyHandling.htm)

**Comment by steveharter** [comment]
@JamesNK on "Ignore" - is this mode necessary and what scenarios would prefer that over "Preserve"?

"Ignore" has potnential inconsistentcy w.r.t. what is serialized vs. what is ignored which would make me question using it in certain scenarios including:
- Serialization order currently depends on reflection order which is not deterministic, so if property `A` and `B` both serialize the same reference, sometimes `A` will be serialized and sometimes `B` will be serialized.
- The root of the graph may change causing a given type to sometimes serialize a reference, and other times not.

In many scenarios the [JsonIgnore] could be used instead to prevent cycles which has the advantage of being deterministic.

However if:
- "Throws" isn't desired
- and "Preserve" isn't desired because the metadata ($id $ref) or other reasons (performance?)
- and [JsonIgnore] can't be used (perhaps attributes can't be used)
- and potential non-deterministic behavior is OK

 then "Ignore" makes sense.

**Comment by terrajobst** [comment]
We seem to lean towards not having `Ignore` -- it results in payloads that nobody can reason about. The right fix for those scenarios to modify the C# types to exclude back pointers from serialization.

We'd like to see more detailed spec for the behavior, specifically error cases.

**Comment by JamesNK** [comment]
Ok, I don't feel strongly that ignore is necessary. And it should never be the default setting so you can always add it in the future if desired.

**Comment by paulovila** [comment]
Hi @Jozkee , is there a way to try PreserveReferencesHandling, like in a nuget preview?

**Comment by Jozkee** [comment]
@paulovila this feature is still a work in progress, once is finished and merged into our master branch you can try it out by [installing the latest bits](https://github.com/dotnet/core-sdk#installers-and-binaries) of Net core.

**Comment by los93sol** [comment]
This is extremely disappointing, it was called out by numerous people while 3.0 was in preview as a complete blocker for a lot of people to use the new serializer and benefit from it.  It’s even more disappointing this still isn’t being prioritized properly.  It seems like by the time this gets addressed people will have just abandoned the new serializer altogether and likely won’t return to it.

**Comment by ahsonkhan** [comment]
> It’s even more disappointing this still isn’t being prioritized properly.

What gave you that impression? We are actively working on adding this feature for 5.0.

**Comment by los93sol** [comment]
IMO this is a major breaking change and it actually roadblocks a lot of people from following the recommended serializer setup from the docs.  I’m frustrated because despite this being called out in preview there’s no plan to address in 3.0 or even 3.1 that I’ve seen.  Waiting for 5.0 is way too far out on the roadmap and why I said it still isn’t being prioritized properly.

**Comment by mythrz** [comment]
I little tdd feedback from using System.Text.Json. I apologise for not being able to get deeper into the "why" or if this is already known and irrelevant, but I think this may be significant to some of us. It is at least a "workaround" for some scenarios...

The object cycle error only triggers when the class is in a different namespace. It appears to have something more besides the object depth... And more about how the System.Text.Json deals with the referenced class to serialise. The right side, shows a classTest working with all the code in the same .cs file. If I used the class from another namespace, it gives the error.

![image](https://user-images.githubusercontent.com/44119535/67282218-29e52e00-f4c9-11e9-80bd-43ec894bee9c.png)

**Workaround (ish) :\ :**
Simply create the class inside the same file (.cs file). Don't use different namespace or files, this way it works.
Also, make sure you are using get and set, else it won't serialise the content, the structure will be empty.
![image](https://user-images.githubusercontent.com/44119535/67282548-ec34d500-f4c9-11e9-8f43-44263ee82dd4.png)

Example when you don't use get and  set props.:
![image](https://user-images.githubusercontent.com/44119535/67282786-636a6900-f4ca-11e9-9ea2-5e0d82f7fdbc.png)

This was the only workaround I found. Indicating the issue may be in the depth of the reference, and not the "depth of the class it self". 
Another working example (both classes and this code are in the same .cs file):
![image](https://user-images.githubusercontent.com/44119535/67288417-db3d9100-f4d4-11e9-91ed-dcec2f8a186c.png)

**Comment by mythrz** [comment]
Is there a way to make the referenced class as if it was in the same .cs file as the System.Text.Json library, virtually? This could be a temporary fix....

**Comment by mythrz** [comment]
Edit: Deleted the other comments. Are meaningless!! I was following the wrong crumbles...
In my case, I had an additional Method (garbage/test method...) inside the class that I wanted to serialize... mb, the essential is working.

**Comment by doogdeb** [comment]
I also experience an issue when trying to serialize the following.

     public class CacheModel
     {
        public Type DtoModelType { get; set; }

        public string CacheKey { get; set; }
      }

DtoModelType is never of type CacheModel but I still get an "JsonException: A possible object cycle was detected which is not supported" error.

**Comment by doogdeb** [comment]
> > It’s even more disappointing this still isn’t being prioritized properly.
> 
> What gave you that impression? We are actively working on adding this feature for 5.0.

Isn't 5.0 a year away?

**Comment by mythrz** [comment]
> > > It’s even more disappointing this still isn’t being prioritized properly.
> > 
> > 
> > What gave you that impression? We are actively working on adding this feature for 5.0.
> 
> Isn't 5.0 a year away?

No. We are currently in 3.0/3.1! Next year, around this time, .NET 5 will be released. They are gonna skip a number (the number 4) so we don't get confused with  .NET Framework 4.*

Next year, there will be one .NET to "rule us all"

**Comment by doogdeb** [comment]
> > > > It’s even more disappointing this still isn’t being prioritized properly.
> > > 
> > > 
> > > What gave you that impression? We are actively working on adding this feature for 5.0.
> > 
> > 
> > Isn't 5.0 a year away?
> 
> No. We are currently in 3.0/3.1! Next year, around this time, .NET 5 will be released. They are gonna skip a number (the number 4) so we don't get confused with .NET Framework 4.*
> 
> Next year, there will be one .NET to "rule us all"

I understand that, but you mention that this change is earmarked for 5.0, which I'm assuming is .Net 5.0, which is a year away.

**Comment by ahsonkhan** [comment]
> I also experience an issue when trying to serialize the following.
> 
> ```
>  public class CacheModel
>  {
>     public Type DtoModelType { get; set; }
> 
>     public string CacheKey { get; set; }
>   }
> ```
> 
> DtoModelType is never of type CacheModel but I still get an "JsonException: A possible object cycle was detected which is not supported" error.

The issue you are seeing is specifically about `System.Type` support and unrelated to cycles/referenceloophandling.

I have filed an issue for this with a potential workaround using converters: https://github.com/dotnet/corefx/issues/42712
Let's move the discussion there.

**Comment by nickcoad** [comment]
> > > > It’s even more disappointing this still isn’t being prioritized properly.
> > > 
> > > 
> > > What gave you that impression? We are actively working on adding this feature for 5.0.
> > 
> > 
> > Isn't 5.0 a year away?
> 
> No. We are currently in 3.0/3.1! Next year, around this time, .NET 5 will be released. They are gonna skip a number (the number 4) so we don't get confused with .NET Framework 4.*
> 
> Next year, there will be one .NET to "rule us all"

Next year, around this time, is just another way of saying "a year away"...

**Comment by ahsonkhan** [comment]
> Isn't 5.0 a year away?

> I understand that, but you mention that this change is earmarked for 5.0, which I'm assuming is .Net 5.0, which is a year away.

> "Next year, around this time" is just another way of saying "a year away"...

Yes, this feature is for 5.0 (i.e. master), which is planned to ship near end of next year (hence the milestone). 3.1 is already done. That said, S.T.Json also ships as a NuGet package so once the feature is done, you could consider using a preview package on 3.1 before 5.0 officially ships.

**Comment by paulovila** [comment]
It's complicated to depend on the newtonsoft serialisation patch because It is not being updated at the same pace as the SignalR patch, and that causes that we can’t upgrade tho the intermediate versions that you provide. 🤷‍♂️

**Comment by ahsonkhan** [comment]
@paulovila - I am sorry, but I don't understand what you mean. What target framework are you using in your project?
- What do you mean by "the newtonsoft serialisation patch"?
- What do you mean by "SignalR patch"?

> and that causes that we can’t upgrade to the intermediate versions that you provide.

Let's say you have a .NET Core/Asp.Net Core 3.1 application. The built-in `System.Text.Json` library that shipped with 3.1 doesn't contain this reference handling feature. However, if you are fine with referencing preview packages, you can add a `PackageReference` to the latest 5.0 nightly build of the `System.Text.Json` NuGet package and get that feature (once its implemented/merged into master).

**Comment by paulovila** [comment]
@ahsonkhan the aforementioned patches are:
```xml
<PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.0.0" />
<PackageReference Include="Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson" Version="3.0.0" />
```
with that you can circumvent the circular references as follows:
```csharp
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json;
...
services.AddMvc()
    .AddNewtonsoftJson(options =>
    {
        options.SerializerSettings.NullValueHandling = NullValueHandling.Ignore;
        options.SerializerSettings.PreserveReferencesHandling = PreserveReferencesHandling.Objects;
    })
    .SetCompatibilityVersion(CompatibilityVersion.Version_3_0);

services.AddSignalR().AddNewtonsoftJsonProtocol(options =>
{
    options.PayloadSerializerSettings.NullValueHandling = NullValueHandling.Ignore;
    options.PayloadSerializerSettings.PreserveReferencesHandling = PreserveReferencesHandling.Objects;
});
```
Is this patch strategy the right way to go? I can see now that there has been some alignment in the nuget preview versions since I tried version 3.0.0.0.

**Comment by Allann** [comment]
Also affected by this issue.  Waiting for a year to be able to use EF in an API/Web app is not an option, will be in production then and a complete change in serialization methods at that stage would carry too much risk.  Therefore the only option is to user NewtonSoft and not the System.Text.Json.

I can't fin how to circumvent the CR as shown above in an Api controller since it doesn't use AddMvc anymore.

Is there any other options?  Maybe the way navigation properties are declared?  Time for some additional testing.

**Comment by paulovila** [comment]
@Allann maybe take a look at this https://github.com/paulovila/UnoPlatformBravent/blob/master/src/UnoWebApiSwagger.WebApi/Startup.cs

**Comment by Allann** [comment]
@paulovila  thanks for the link, but try upgrading that to 3.1 and see how you go ;)  AddMvc is not part of the recommended templated code for a WebAPI anymore.  Instead they recommend AddControllers.
For example the `ConfigureServices` is this:
``` csharp
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddDbContext<CountryContext>(opt =>
               opt.UseInMemoryDatabase("Countries"));
            services.AddControllers();
        }
```

This works fine and produces an very fast and responsive RESTful service.  I have managed to get the required request working by changing the navigation properties, and the way in which I created the resultset.  Removing the Include was the first step, but a pretty crappy work-around, as I now need 3 different models to provide the same functionality, just because `Include` get's "fixed" when I don't want it in the first place.

**Comment by paulovila** [comment]
This is breaking the continuity of netcore 3.0 to 3.1. In order to work 3.1, I think you guys should fix the "Microsoft.AspNetCore.Mvc.NewtonsoftJson" and "Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson" nuget patches for the time being

**Comment by BrennanConroy** [comment]
> should fix the "Microsoft.AspNetCore.Mvc.NewtonsoftJson" and "Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson"

What needs to be fixed with these packages? If you have found a problem in those packages please file an issue at https://github.com/aspnet/AspNetCore/issues/new/choose

**Comment by paulovila** [comment]
@Allann, @BrennanConroy  I can see now that patches are working with NetCore 3.1, I think there was some intermediate version in ```Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson``` that had a refence colliding ```Microsoft.AspNetCore.Mvc.NewtonsoftJson```, so no need to report anything :)

**Comment by Arash-Sabet** [comment]
We had to change the code below:
```
var domainObject = payload.GetMyObject();
parentObject.DomainObject = domainObject;
... // ommitted code here
DoSomething(parentObject.DomainObject);
updatedPayload = JsonSerializer.Serialize(parentObject.DomainObject);
```
to the following one to avoid circular reference exception:
```
var domainObject = payload.GetMyObject();
... // ommitted code here
DoSomething(parentObject.DomainObject);
updatedPayload = JsonSerializer.Serialize(domainObject);
```
Is it related to the case that @Jozkee is suggesting to fix/improve?  Not sure why the 2nd code snippet does not throw that exception while both domainObject and parentObject.DomainObject are pointing to the same object?

**Comment by Jozkee** [comment]
@Arash-Sabet I can see how that can cause an infinite loop if `domainObject` contains a reference to `parentObject` somewhere.

**Comment by Arash-Sabet** [comment]
@Jozkee For sure ```domainObject``` has no reference to its ```parentObject``` at all.

**Comment by ahsonkhan** [comment]
> For sure ```domainObject``` has no reference to its ```parentObject``` at all.

Can you share what the type and POCO structure of domainObject and parentObject looks like? A simplified repro would also help to address your concern.

> Not sure why the 2nd code snippet does not throw that exception while both domainObject and parentObject.DomainObject are pointing to the same object?

I can't tell if that's true just from the snippet provided. How do you know that `domainObject and parentObject.DomainObject are pointing to the same object`?

**Comment by etghanibahman** [comment]
Do you have any advice on how we can serialize DataSet with current version of System.Text.Json.JsonSerializer ?

**Comment by drdamour** [comment]
man, this thing is not ready for very simple use cases....

**Comment by Jozkee** [comment]
@drdamour Thank you for give it a try.
Regardless of https://github.com/dotnet/runtime/pull/655 being merged, it is still a WIP. 
Can you specify the case that is causing you issues?

**Comment by drdamour** [comment]
simple double relationship from db office <=> employee getting either an including the other circles back to the other now on deserialization.  can't mark them as do not serialize because i need to be able to get from either way.

**Comment by Jozkee** [comment]
@drdamour You mean `Office` and `Employee` enitites where `Office` contains a `List<Employee>`? and each `Employee` contains an `Office` property?

Can you provide your model classes in a comment?

**Comment by Jozkee** [comment]
Forgot to add the comment here.

> The preview version for `ReferenceHandling` feature is here.
> 
> * Download the `System.Text.Json` preview package version `5.0.0-alpha.1.20071.1` from dotnet5 gallery (through Visual Studio's NuGet client):
> https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet5/nuget/v3/index.json
> 
> * How to use it:
> Set `ReferenceHandling.Preserve` to `ReferenceHandling` property in `JsonSerializerOptions`.
> ```cs
> var options = new JsonSerializerOptions
> {
>     ReferenceHandling = ReferenceHandling.Preserve
> };
> 
> string json = JsonSerializer.Serialize(objectWithLoops, options);
> ```
> Here's the [spec doc](https://github.com/dotnet/runtime/blob/c23959daba1ccc0f27d74deecf486625304cde7e/src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md) including samples and notes about many scenarios.
> If you want to add feedback, feel free to leave a comment.

**Comment by Dzivo** [comment]
@Jozkee  is this available in 3.1.1 version ?

**Comment by Jozkee** [comment]
@Dzivo  No, this is going to be shipped on 5.0; right now it is only available on the preview package version of `System.Text.Json`.

**Comment by Dzivo** [comment]
@Jozkee thx for the info switching back to NewtonsoftJson until then :)
